### PR TITLE
Support `on_schema_change` feature on incremental model

### PR DESCRIPTION
**Pull request by hiro-o918**
## Problem
Currently, dbt-athena does not support `on_schema_change` option.

## Solution

- Add custom column class `AthenaColumn`
  - Athena has different column type between DML and DDL (e.g. integer)
    - https://docs.aws.amazon.com/athena/latest/ug/data-types.html
  - To avoid error on `alter table` statement because of difference of column type
- Overwrite macro `sync_column_schemas`
  - Athena does not support `drop columns` statement, but `replace columns`
    - https://docs.aws.amazon.com/athena/latest/ug/alter-table-replace-columns.html
  - this causes inconsistency with original `sync_column_schemas`, so overwrite this macro on this adaptor

## Note
Just cherry-pick the commit of these PRs(https://github.com/dbt-athena/dbt-athena/pull/3, https://github.com/dbt-athena/dbt-athena/pull/28)

### CODE REVIEW

File: dbt/adapters/athena/__init__.py [hunk 1]
```diff
@@ -1,10 +1,16 @@
 from dbt.adapters.base import AdapterPlugin
 
-from dbt.adapters.athena.connections import (  # NOQA
-    AthenaConnectionManager,
-    AthenaCredentials,
-)
+from dbt.adapters.athena.column import AthenaColumn
```
Comment (hiro-o918) [inline dbt/adapters/athena/__init__.py:3]: fixed by pre-commit and isort

File: dbt/adapters/athena/column.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
+from dbt.adapters.base import Column as BaseColumn
+
+
+class AthenaColumn(BaseColumn):
```
Comment (nicor88) [inline dbt/adapters/athena/column.py:4]: Did you consider to use a macro instead?

File: dbt/adapters/athena/column.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    @property
+    def data_type(self) -> str:
+        # NOTE: Athena has different types between DML and DDL, use DDL type in this case
+        #       ref: https://docs.aws.amazon.com/athena/latest/ug/data-types.html
+        if self.dtype.lower() == "integer":
+            return "int"
+        elif self.is_string():
+            return self.string_type(self.string_size())
```
Comment (nicor88) [inline dbt/adapters/athena/column.py:12]: Doesn't athena pretend to have always strings?

File: dbt/adapters/athena/column.py (hunk 2 above)
Comment (hiro-o918) [reply dbt/adapters/athena/column.py:4]: Yes, It's ok to move implementation of this class to macro.
should I wait for #26 merged?

File: dbt/include/athena/macros/materializations/models/incremental/incremental.sql [hunk 4]
```diff
@@ -10,6 +10,7 @@
 
   {% set raw_strategy = config.get('incremental_strategy', default='insert_overwrite') %}
   {% set strategy = validate_get_incremental_strategy(raw_strategy) %}
+  {% set on_schema_change = incremental_validate_on_schema_change(config.get('on_schema_change'), default='ignore') %}
```
[the remark this answers is not in the document]
Comment (hiro-o918) [reply dbt/include/athena/macros/materializations/models/incremental/incremental.sql:13]: It comes from dbt-core implementation.
I'll do update REAMDE about this.

https://github.com/dbt-labs/dbt-core/blob/7efb6ab62dae3ea9da4e6e28e9a30aa4e0e74919/core/dbt/include/global_project/macros/materializations/models/incremental/on_schema_change.sql#L1-L16

File: dbt/adapters/athena/column.py (hunk 3 above)
Comment (hiro-o918) [reply dbt/adapters/athena/column.py:12]: Athena supports `string` and `varchar(n)`, so I choose this implementation, but I think `string` is clearer than `varchar(n)`, so I'll replace this whit `string`

image

**Comment by hiro-o918** [comment]
By the way, dbt may be updated to add a comment about metadata of environment to query,
which causes error on DDL (not DML) 

This is an example of this.

```sql
/* {"app": "dbt", "dbt_version": "1.3.0", "profile_name": "test", "target_name": "dev", "node_id": "model.foo.partitioned_model"} */

    alter table dbt_sandbox.partitioned_model
          replace columns (id int, additional_column_1 string, additional_column_2 varchar(3))
```

So, I'll check the way to avoid adding such a comment.

File: dbt/adapters/athena/column.py (hunk 3 above)
Comment (nicor88) [reply dbt/adapters/athena/column.py:12]: as we would like to use the same approach for iceberg, and there varchar is not supported, also IMHO we should stick with string.

File: dbt/include/athena/macros/materializations/models/incremental/incremental.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/incremental/incremental.sql:13]: Nice thanks, indeed some notes in the Readme will be nicer.

File: dbt/adapters/athena/column.py (hunk 2 above)
Comment (nicor88) [reply dbt/adapters/athena/column.py:4]: I think that the #26 is slightly different. You might want to build something like that: https://github.com/dbt-athena/dbt-athena/blob/main/dbt/include/athena/macros/materializations/models/helpers_iceberg.sql#L102-L124 
so feel free to proceed.

File: dbt/include/athena/macros/materializations/models/incremental/incremental.sql (hunk 4 above)
[the remark this answers is not in the document]
Comment (hiro-o918) [reply dbt/include/athena/macros/materializations/models/incremental/incremental.sql:13]: [7ae9225](https://github.com/dbt-athena/dbt-athena/pull/30/commits/7ae9225b25aada0ee41c3bea3bc9db378f3010fc)
Updated README.

File: dbt/adapters/athena/column.py (hunk 2 above)
Comment (hiro-o918) [reply dbt/adapters/athena/column.py:4]: `helpers_iceberg.sql ` was very helpfull, I move the AthenaColumn implementation to  macro

[46c0121](https://github.com/dbt-athena/dbt-athena/pull/30/commits/46c012158cdcfcdaa724ac4b5e636d5808d803d6)

**Comment by hiro-o918** [comment]
https://github.com/dbt-athena/dbt-athena/pull/30#issuecomment-1321163344

The above is resolved by [ee0e69c](https://github.com/dbt-athena/dbt-athena/pull/30/commits/ee0e69c015a35d5e8e01608acc5f1db1d3597119)

File: dbt/include/athena/macros/materializations/models/incremental/helpers.sql [hunk 5]
```diff
@@ -54,3 +57,37 @@
...
+      {%- do columns.append(column) -%}
+    {%- endif -%}
+  {%- endfor -%}
+  {{ return(columns) }}
+{%- endmacro %}
+
+{% macro ddl_data_type(col_type) -%}
+    -- transform varchar
+  {% set re = modules.re %}
+  {% set ns = namespace(ddl_type=col_type) %}
```
Comment (Jrmyy) [inline dbt/include/athena/macros/materializations/models/incremental/helpers.sql:74]: We need to be careful, we have two identical implementations of the same thing, and we just changed it yesterday on the iceberg helpers, we might want to factorize this a little.

File: dbt/adapters/athena/impl.py [hunk 6]
```diff
@@ -26,6 +27,7 @@
 class AthenaAdapter(SQLAdapter):
     ConnectionManager = AthenaConnectionManager
     Relation = AthenaRelation
+    Column = AthenaColumn
```
Comment (Jrmyy) [inline dbt/adapters/athena/impl.py:30]: Do we still need this or is it covered by the macros ?

File: dbt/include/athena/macros/materializations/models/incremental/helpers.sql (hunk 5 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/incremental/helpers.sql:74]: if the implementation is identical, let's move it under  models/utils/ddl_data_type.sql
and then use that in iceberg too.

File: dbt/adapters/athena/impl.py (hunk 6 above)
Comment (nicor88) [reply dbt/adapters/athena/impl.py:30]: Should be possible to drop that.

File: dbt/include/athena/macros/materializations/models/incremental/helpers.sql (hunk 5 above)
Comment (hiro-o918) [reply dbt/include/athena/macros/materializations/models/incremental/helpers.sql:74]: I think so.
Can I move this to that place and commonoze it?

File: dbt/adapters/athena/impl.py (hunk 6 above)
Comment (hiro-o918) [reply dbt/adapters/athena/impl.py:30]: Oops, sorry.
Thanks 🙏

File: dbt/include/athena/macros/materializations/models/incremental/helpers.sql (hunk 5 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/incremental/helpers.sql:74]: Absolutely, that was the suggestion @Jrmyy , and picking up from there I suggest you to:
* move the macro into: `models/utils/ddl_data_type.sql`
* refactor iceberg to use that new macro
* use the new macro on this on_schema change

File: dbt/adapters/athena/impl.py (hunk 6 above)
Comment (hiro-o918) [reply dbt/adapters/athena/impl.py]: removed by [ca0c45a](https://github.com/dbt-athena/dbt-athena/pull/30/commits/ca0c45abf05a84e9708edc24e232a79ab10c48bd)

File: dbt/include/athena/macros/utils/ddl_data_type.sql [hunk 7]
```diff
@@ -0,0 +1,25 @@
+{# Athena has different types between DML and DDL #}
+{# ref: https://docs.aws.amazon.com/athena/latest/ug/data-types.html #}
+{% macro ddl_data_type(col_type) -%}
+    -- transform varchar
+  {% set re = modules.re %}
+  {% set ns = namespace(ddl_type=col_type) %}
+  {% set matching_strings = re.findall('(varchar\(\d+\)|character varying\(\d+\))', ns.ddl_type) %}
```
Comment (Jrmyy) [inline dbt/include/athena/macros/utils/ddl_data_type.sql:7]: Why did you remove the re.sub implementation ? 
Furthermore, it seems that this implementation is not covering an issue we had about varchar without length.

This should work.
```
{% set data_type = re.sub('(?:varchar|character varying)(?:\(\d+\))?', 'string',  col_type) %}
```

**Review by Jrmyy — CHANGES_REQUESTED** [review]

File: dbt/include/athena/macros/utils/ddl_data_type.sql (hunk 7 above)
Comment (Jrmyy) [reply dbt/include/athena/macros/utils/ddl_data_type.sql:7]:
```python
>>> x = 'map(map(character varying(22), varchar), map(character varying, varchar(244)))'
>>> re.sub('(?:varchar|character varying)(?:\(\d+\))?', 'string',  x)
'map(map(string, string), map(string, string))'
```

Comment (hiro-o918) [reply dbt/include/athena/macros/utils/ddl_data_type.sql]: Thank you, I forget use updated version of iceberg_helper :bow:
fixed [fe25801](https://github.com/dbt-athena/dbt-athena/pull/30/commits/fe25801da1018af106b5acb747f5825eda72ae84)

File: dbt/include/athena/macros/materializations/models/incremental/helpers.sql (hunk 5 above)
Comment (nicor88) [reply dbt/include/athena/macros/materializations/models/incremental/helpers.sql]: We could add another little section to replace character with an empty value, and then let ddl type handling do the rest as now.

**Review by Jrmyy — APPROVED** [review]

**Review by nicor88 — APPROVED** [review]
Nice job.

**Comment by hiro-o918** [comment]
Thank you for your quick reviewing!
